pkg/operator/csi/csidefaultstorageclasscontroller/csi_default_storageclass_controller.go

```diff
@@ -0,0 +1,135 @@
+// Also, the controller will emit alerts if more than one StorageClass defined
+// as default exists - this is not a recommended setup.
+// It produces following Conditions:
+// DefaultStorageClassControllerDegraded - failed to apply StorageClass provided
+type CSIDefaultStorageClassController struct {
+	name               string
+	assetFunc          resourceapply.AssetFunc
+	file               string
+	kubeClient         kubernetes.Interface
+	storageClassLister v1.StorageClassLister
+	operatorClient     v1helpers.OperatorClient
+	eventRecorder      events.Recorder
+}
+
+func NewCSIDefaultStorageClassController(
+	name string,
+	assetFunc resourceapply.AssetFunc,
+	file string,
+	kubeClient kubernetes.Interface,
+	InformerFactory informers.SharedInformerFactory,
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &CSIDefaultStorageClassController{
+		name:               name,
+		assetFunc:          assetFunc,
+		file:               file,
+		kubeClient:         kubeClient,
+		storageClassLister: InformerFactory.Storage().V1().StorageClasses().Lister(),
+		operatorClient:     operatorClient,
+		eventRecorder:      eventRecorder,
+	}
+
+	return factory.New().WithSync(
+		c.Sync,
+	).ResyncEvery(
+		time.Minute,
+	).WithSyncDegradedOnError(
+		operatorClient,
+	).WithInformers(
+		operatorClient.Informer(),
+		InformerFactory.Storage().V1().StorageClasses().Informer(),
+	).ToController(
+		"DefaultStorageClassController",
+		eventRecorder,
+	)
+}
+
+func (c *CSIDefaultStorageClassController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("DefaultStorageClassController sync started")
+	defer klog.V(4).Infof("DefaultStorageClassController sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	syncErr := c.syncStorageClass(ctx)
+
+	return syncErr
+}
+
+func (c *CSIDefaultStorageClassController) syncStorageClass(ctx context.Context) error {
+	expectedScBytes, err := c.assetFunc(c.file)
+	if err != nil {
+		return err
+	}
+
+	expectedSC := resourceread.ReadStorageClassV1OrDie(expectedScBytes)
+
+	existingSCs, err := c.storageClassLister.List(labels.Everything())
+	if err != nil {
+		klog.V(2).Infof("could not list StorageClass objects")
+		return err
+	}
+
+	defaultScCount := 0
+	expectedScNonDefaultExists := false
+	for _, sc := range existingSCs {
+		if sc.Annotations[defaultScAnnotationKey] == "true" && sc.Name != expectedSC.Name {
+			defaultScCount++
+		}
+		if sc.Annotations[defaultScAnnotationKey] == "false" && sc.Name == expectedSC.Name {
+			expectedScNonDefaultExists = true
+		}
+	}
+
+	// Change expected SC to non-default if there already is a default or expected SC is already present as non-default.
+	if defaultScCount > 0 || expectedScNonDefaultExists {
```

Comment:
I still think it's important to distinguish if `expectedSC` already exists or not. If it exists and it's default, its annotation will be set to `false` when user creates another default SC. Which is quite surprising.

In other word, if `expectedSC` exists -> don't touch its annotations (i.e. copy `is-default-class` from the existing one). 
If `expectedSC` does not exist, adjust its default annotation as below and create it.

Reply:
Good point, changed the code a bit and added another example test in the description to demonstrate this.